src/param.rs

```diff
@@ -28,28 +30,38 @@ pub struct AudioParamOptions {
     pub max_value: f32,
 }
 
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+#[derive(PartialEq, Eq, Debug)]
 enum AudioParamEventType {
     SetValue,
     SetValueAtTime,
     LinearRampToValueAtTime,
     ExponentialRampToValueAtTime,
     CancelScheduledValues,
     SetTargetAtTime,
-    // CancelAndHoldAtTime,
-    // SetValueCurve,
+    CancelAndHoldAtTime,
+    SetValueCurveAtTime,
 }
 
-#[derive(Copy, Clone, Debug)]
+#[derive(Debug)]
 pub(crate) struct AudioParamEvent {
     event_type: AudioParamEventType,
     value: f32,
     time: f64,
-    time_constant: Option<f64>,
+    time_constant: Option<f64>, // populated by `SetTargetAtTime` events
+    cancel_time: Option<f64>,   // populated by `CancelAndHoldAtTime` events
+    duration: Option<f64>,      // populated by `SetValueCurveAtTime` events
+    values: Option<Box<[f32]>>, // populated by `SetValueCurveAtTime` events
 }
 
-// Event queue that mimics the `BinaryHeap` API while allowing `retain` without going nigthly
-// @note - diverge in that `queue.sort()` is stable and must be called explicitly
+// Event queue that contains `AudioParamEvent`s, most of the time, events must be
+// ordered (using stable sort), some operation may break this ordering (e.g. `push`)
+// in which cases `sort` must be called explicitely.
+// In the current implementation of the param rendering, `sort` is called once after
+// all events have been inserted in the queue at each `tick` (with the exception
+// `CancelAndHoldAtTime` which needs a clean queue to find its neighbors, but this
+// occurs during the insertion of events)
+// After this point, the queue should be considered sorted and no operations that
+// breaks the ordering should be done.
 #[derive(Debug)]
 struct AudioParamEventTimeline {
```

Comment:
I have some suggestions to put the `sorted/dirty` state into the type system here (to avoid panics and avoid duplicate sorts) but I will just put them in a PR later. The timeline is surely fine for now

Reply:
ok cool